server/src/main/java/org/elasticsearch/common/geo/DimensionalShapeType.java

```diff
@@ -0,0 +1,180 @@
+            public DimensionalShapeType visit(Circle circle) {
+                throw new IllegalArgumentException("invalid shape type found [Circle] while computing dimensional shape type");
+            }
+
+            @Override
+            public DimensionalShapeType visit(Line line) {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.LINESTRING);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(LinearRing ring)  {
+                throw new UnsupportedOperationException("should not visit LinearRing");
+            }
+
+            @Override
+            public DimensionalShapeType visit(MultiLine multiLine)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.MULTILINESTRING);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(MultiPoint multiPoint)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.MULTIPOINT);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(MultiPolygon multiPolygon)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.MULTIPOLYGON);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(Point point)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.POINT);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(Polygon polygon)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.POLYGON);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(Rectangle rectangle)  {
+                st = DimensionalShapeType.max(st, DimensionalShapeType.POLYGON);
+                return st;
+            }
+
+            @Override
+            public DimensionalShapeType visit(GeometryCollection<?> collection)  {
+                for (Geometry shape : collection) {
+                    shape.visit(this);
+                }
+                int dimension = st.centroidDimension();
+                if (dimension == 0) {
+                    return DimensionalShapeType.GEOMETRYCOLLECTION_POINTS;
+                } else if (dimension == 1) {
+                    return DimensionalShapeType.GEOMETRYCOLLECTION_LINES;
+                } else {
+                    return DimensionalShapeType.GEOMETRYCOLLECTION_POLYGONS;
+                }
+            }
+        });
+    }
+
+    /**
+     * The integer representation of the dimension for the specific
+     * dimensional shape type. This is to be used by the centroid
+     * calculation to determine whether to add a sub-shape's centroid
+     * to the overall shape calculation.
+     *
+     * @return 0 for points, 1 for lines, 2 for polygons
+     */
+    private int centroidDimension() {
```

Comment:
Please pardon the drive-by commentary, but why not just make `centroidDimension` a field on the enum?  That guards against later adding a new entry and forgetting to update this function, for example.

Reply:
thanks Mark. due to a bug I found when indexing polygons that degrade to lines, or lines that degrade to points. the ability to distinguish all the various shapes is to be removed. opened #52500 to clean this up. the whole `centroidDimension` is to be removed